keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +336,169 @@ or later. The dual-stack feature requires no changes to this API.
 The versions of CNI plugin binaries that must be used for proper dual-stack functionality (and IPv6 functionality in general) depend upon the version of the container runtime that is used in the cluster nodes (see [CNI issue #531](https://github.com/containernetworking/cni/issues/531) and [CNI plugins PR #113](https://github.com/containernetworking/plugins/pull/113)):
 
 ### Services
+Services depend on `--services-cluster-ip-range` for `ClusterIP` assignment. A dualstack cluster can have this flag configured as `--services-cluster-ip-rang=<cidr>,<cidr>` or `--services-cluster-ip-rang=<cidr>`. The following rules apply:
+
+1. The maximum allowed number of cidrs in this flag is two. Attempting to use more than two cidrs will generate failures during startup. 
+2. When configured with two cidrs then the values are expected to have different ip families. e.g `--services-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>` or `--services-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`. Attempting to configure a cluster with two similar cidrs will generate failures during startup.
+3. A dualstack cluster can have a single cidr in this flag. That means the cluster will be to egress traffic on both ip families, but will *not* be ingress traffic on both ip families.
+4. When upgrading a cluster to dualstack, the first element of the `--services-cluster-ip-range` flag MUST NOT change IP families.  As an example, an IPv4 cluster can safely be upgraded to use `--services-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>`, but can not safely be changed to `--services-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`.
+5. The first cidr entry of `--services-cluster-ip-range=<first cidr entry>, <secondary cidr entry>` is considered cluster default address family.
+6. Attempting to allocate `ClusterIP` from an ip address family that is not configured on the cluster will result in an error.
+
+> The below discussion assumes that `Endpoint Slice and Endpoint Slice Controller`  will allow dual stack endpoints, the existing endpoint controller will remain single stack. That means enabling dual stack services by having more than one entry in `--services-cluster-ip-range` without enabling `Endpoint Slice` feature gate will result in validation failures during startup.
```

Comment:
They just use the first IP family for `Endpoints`

Reply:
There will be validation to ensure that len(cidr-list) + 1 && `EndPointSlice Feature is enabled`.

Reply:
@khenidak I don't think you and @thockin are saying the same thing.

It seems to me that you are saying that if you run `kube-apiserver --service-cluster-ip-ranges 172.30.0.0/16,fd02::/112 --feature-gates EndpointSlice=false` then kube-apiserver will exit with an error.

It seems to me that Tim is saying that you are allowed to do that, but that EndpointController will only ever create single-stack Endpoints. I guess it's not actually 100% clear what he was saying. I was interpreting it as "For each Service, EndpointController will create Endpoints containing IPs of the same family as the Service's `clusterIPs[0]`" but maybe he meant "For all Services, EndpointController will create Endpoints containing IPs of the same family as the first argument to `--service-cluster-ip-ranges`"?

Reply:
I think I answered the wrong question.

If flags config says dual-stack but EPSlice is not gated:

On the one hand there's very little reason not to enable EPSlice at this point. On the other hand, this should be valid, I think

Any given Service's Endpoints might be one family or the other.  Without EPSlice there's no way to represent "both", but that's really orthogonal.

Reply:
What we are proposing in the KEP is: if you choose to have dual stack cidrs, then you must have EndPointSlice enabled. So `kube-apiserver --service-cluster-ip-ranges 172.30.0.0/16,fd02::/112 --feature-gates EndpointSlice=false` must fail api-server boot-up sequence even if the user have no `dual stack` services, because the system will automatically create dual stack default services. And more importantly we don't know at `validate service step` if EndPointSlice is enabled or not.

The other point, I think @thockin covered.